EIPS/eip-2315.md

```diff
@@ -0,0 +1,91 @@
+---
+eip: 2315
+title: Simple Subroutines for the EVM
+status: Draft
+type: Standards Track
+category: Core
+author: Greg Colvin
+discussions-to: https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm/3941
+created: 2019-10-17
+---
+## Abstract
+
+This proposal introduces two opcodes to support subroutines:  `JUMPSUB` and `RETSUB`.
+
+## Motivation
+
+The EVM does not provide subroutines as a primitive.  Instead, calls must be synthesized by fetching and pushing the current program counter on the data stack and jumping to the subroutine address; returns must be synthesized by getting the return address to the top of stack and jumping back to it.
+
+## Specification
+
+**_JUMPSUB_**
+
+* Jumps to the address on top of the stack.  This must be a `JUMPDEST`.
+
+**_RETSUB_**
+
+* Returns to the most recently executed `JUMPSUB` instruction.
+
+## Rationale
+
+This is the smallest possible change that provides native subroutines without breaking backwards compatibility.  It does not enforce validity like [EIP-615](https://eips.ethereum.org/EIPS/eip-615), but code that
+* pushes a constant argument before every `JUMP`, `JUMPI` and `JUMPSUB`
+* ensures that the _data stack_ is the same size at every execution of each `JUMP`, `JUMPI`, `JUMPSUB` or `JUMPDEST` operation.
+
+is valid and will meet all of the [EIP-615](https://eips.ethereum.org/EIPS/eip-615) safety conditions.  In particular, the Yellow paper defines five exceptional halting states.  
+1. Insufficient gas
+2. More than 1024 stack items
+3. Insufficient stack items
+4. Invalid jump destination
+5. Invalid instruction
+
+Valid code cannot violate states 3, 4, or 5.
+
+## Backwards Compatibility
+
+These changes do not affect existing EVM code.
+
+## Test Cases
+```
+PUSH 4     // before jumpsub: PC = 3
+JUMPSUB    // after jumpsub:  PC = 4
+JUMPDEST   // before retsub:  PC = 4
+RETSUB     // after retsub:   PC = 3
```

Comment:
I don't agree with your PC calculation, unless your `PUSH` is a `PUSH2` which doesn't make sense. I suggest explicitly showing the PC and what type of `PUSH` it is instead of typing "before xx", "after xx"
```
00: PUSH1 3
02: JUMPSUB
03: JUMPDEST
04: RETSUB
```

To further clarify, I'd recommend adding a 'trace' aswell, so we see what happens when executing the code aboce. Example: 

```
step, op, stack
0, PUSH1 3 , []
1, JUMPSUB, [3]
2, JUMPDEST, []
3, RETSUB, []
4, JUMPSUB [] ; At this point, it reaches stack-too-shallow, right?
```

Reply:
Agreed.  Do we have a standard syntax for doing this?  This one is fine.

Reply:
This is what I had in mind.
```
step op       stack
0    PUSH1 3  []
1    JUMPSUB  [3]
4    STOP
2    JUMPDEST []
3    RETSUB   []
```

Reply:
Don't confuse `step` with `pc`, they're separate. `step` should be be always increasing. If you provide one `disassembly` with fields `pc, op` and one trace with `step, pc, op, stack` that would make it clearer. 
And I'm wondering about the intent after the last step. Is the intent that `RETSUB` lands on the pc one step after the last `JUMPSUB`?